### DISCUSSION TITLE
Change just a part of a line - zos_lineinfile

### DISCUSSION THREAD

**Discussion by ghost** [ansible-collections/ibm_zos_core#902] [Q&A] [answered]
Hello everybody!
I would like to know if there is a way to edit a PDS member, as zos_lineinfile does, but changing only a part of the line, not being necessary to edit the entire line (as if it were a variable). In the mainframe there are members that have the same information, but not exactly the same spaces or parameters. The line below, for example, may have different spacing between OUTPUT and DD, which would change the search method.

"//OUTPUT DD DSN=LBS.CICP01.V560.SDFHAUTH

I know I can change the regex to find just the value I want, but even then the entire line must be "rewritten". In some cases this can generate errors, delete other parameters, etc. It would be interesting if I could just find V560, for example, and change it to V610, in case of versioning a product.
I've researched and haven't found a way to do exactly that and if it doesn't exist feel free to move this discussion to IDEAS.
In the maiframe world, this will be very useful because a large part of the work is job submission, where you only change one variable, the version of a product, the environment where it runs, the name of the library, etc.

**Comment by ddimatos** [comment, marked as answer]
Hi @fabiobncbb , we don't have a module yet that can perform word replacement, while such a module is in our internal backlog (zos_replace) , the module design is not yet complete there for not in our public backlog, but the good news is that since Ansible is so flexible there are other ways to achieve this. 

My recommendation is to use ZOAU [dsed](https://www.ibm.com/docs/en/zoau/1.2.x) directly. Only note that there is a but we are looking into so you can't perform something like:

```
dsed 's/V[0-9]{3}/V000/g'
``` 

But you can do this instead:
```
command: "dsed 's/V590/V610/g' 'ANSIBLE.LINEFILE.SEQ'"
```

Playbook recreate (i know you mentioned PDSE, but leave that exercise up to you, I used a sequential, should be same):
```

    - name: Create seq data set
      zos_data_set:
        name: ANSIBLE.LINEFILE.SEQ
        type: SEQ
        space_primary: 3
        space_type: M
        record_format: FB
        record_length: 1024
        replace: yes
      register: result

    - debug:
       var: result

    - name: Populate data set
      zos_copy:
        content: |1
          D160882
          D160882 with more info
          ED160882
          This is a line
          //OUTPUT DD DSN=LBS.CICP01.V560.SDFHAUTH
          //OUTPUT DD DSN=LBS.CICP01.V590.SDFHAUTH
          //OUTPUT DD DSN=LBS.CICP01.V590.SDFHAUTHABCDE
          //OUTPUT DD DSN=LBS.CICP01.V560.SDFHAUTH
          //OUTPUT DD DSN=LBS.CICP01.V111.SDFHAUTH
        dest: ANSIBLE.LINEFILE.SEQ
        force: yes
      register: result

    - debug:
       var: result

    - name: Use ZOAU 'dcat' to show data before editing
      command: "dcat 'ANSIBLE.LINEFILE.SEQ'"
      register: result

    - name: Response
      debug:
        msg: "{{ result.stdout_lines }}"

    # Replace V590 with V610 in lines such as '//OUTPUT DD DSN=LBS.CICP01.V590.SDFHAUTH'
    - name: Use ZOAU 'dsed' to edit a SEQ data entry
      command: "dsed 's/V590/V610/g' 'ANSIBLE.LINEFILE.SEQ'"
      register: result

    - name: Response
      debug:
        msg: "{{ result }}"

    - name: Use ZOAU 'dcat' to show data after editing
      command: "dcat 'ANSIBLE.LINEFILE.SEQ'"
      register: result

    - name: Response
      debug:
        msg: "{{ result.stdout_lines }}"
```

Output:
```
TASK [Use ZOAU 'dcat' to show data before editing] 
changed: [zvm]

TASK [Response] 
ok: [zvm] => {
    "msg": [
        " D160882",
        " D160882 with more info",
        " ED160882",
        " This is a line",
        " //OUTPUT DD DSN=LBS.CICP01.V560.SDFHAUTH",
        " //OUTPUT DD DSN=LBS.CICP01.V590.SDFHAUTH",
        " //OUTPUT DD DSN=LBS.CICP01.V590.SDFHAUTHABCDE",
        " //OUTPUT DD DSN=LBS.CICP01.V560.SDFHAUTH",
        " //OUTPUT DD DSN=LBS.CICP01.V111.SDFHAUTH"
    
}

TASK [Use ZOAU 'dsed' to edit a SEQ data entry] 
changed: [zvm]

TASK [Response] 
ok: [zvm] => {
    "msg": {
        "changed": true,
        "cmd": [
            "dsed",
            "s/V590/V610/g",
            "ANSIBLE.LINEFILE.SEQ"
        ],
        "delta": "0:00:00.520654",
        "end": "2023-07-26 17:43:54.757494",
        "failed": false,
        "msg": "",
        "rc": 0,
        "start": "2023-07-26 17:43:54.236840",
        "stderr": "",
        "stderr_lines": [],
        "stdout": "",
        "stdout_lines": []
    }
}

TASK [Use ZOAU 'dcat' to show data after editing] 
changed: [zvm]

TASK [Response] 
ok: [zvm] => {
    "msg": [
        " D160882",
        " D160882 with more info",
        " ED160882",
        " This is a line",
        " //OUTPUT DD DSN=LBS.CICP01.V560.SDFHAUTH",
        " //OUTPUT DD DSN=LBS.CICP01.V610.SDFHAUTH",
        " //OUTPUT DD DSN=LBS.CICP01.V610.SDFHAUTHABCDE",
        " //OUTPUT DD DSN=LBS.CICP01.V560.SDFHAUTH",
        " //OUTPUT DD DSN=LBS.CICP01.V111.SDFHAUTH"
    ]
}
```

I believe this should answer your question and be able to mark this resolved and close. Watch our issues for a zos_replace module, it is not in plan, will be like be an early 2024 item.
